compiler/src/iree/compiler/Codegen/Dialect/PCF/Transforms/Transforms.cpp

```diff
@@ -5,9 +5,268 @@
+    SetVector<Operation *> opsToReplace;
+    if (!result.use_empty() || !isa<RankedTensorType>(result.getType()) ||
+        !isBlockArgDroppable(regionArg, opsToErase, opsToReplace)) {
+      resultsToKeep.push_back(result);
+      newResultTypes.push_back(result.getType());
+      newIsTied.push_back(isTied);
+      if (isTied) {
+        newTiedArgs.push_back(op.getInits()[currTiedIndex]);
+        ++currTiedIndex;
+      } else {
+        int64_t numDynamicDims =
+            cast<ShapedType>(result.getType()).getNumDynamicDims();
+        ValueRange currDynamicSizes =
+            op.getDynamicSizes().slice(currSizesIndex, numDynamicDims);
+        newDynamicSizes.append(currDynamicSizes.begin(),
+                               currDynamicSizes.end());
+        currSizesIndex += numDynamicDims;
+      }
+      continue;
+    }
+
+    // Replace all reads with direct reads from either the tied init or an empty
+    // initial value.
+    if (failed(replaceReadSliceOps(rewriter, op, regionArg, opsToReplace))) {
+      return failure();
+    }
+
+    // Iterate the ops to erase in reverse to make sure ops without users are
+    // erased first.
+    for (auto op : llvm::reverse(opsToErase)) {
+      rewriter.eraseOp(op);
+    }
+
+    blockArgsToDrop.push_back(regionArg.getArgNumber());
+
+    // Increment the dynamic dim/tied operand counters to skip over them.
+    if (isTied) {
+      ++currTiedIndex;
+    } else {
+      currSizesIndex += cast<ShapedType>(result.getType()).getNumDynamicDims();
+    }
+  }
+
+  if (newResultTypes.size() == op->getNumResults()) {
+    return failure();
+  }
+
+  OpTy newOp = cloneWithNewResultTypes(rewriter, op, newResultTypes,
+                                       newTiedArgs, newDynamicSizes, newIsTied);
+
+  // Erase the block arguments associated with the dropped results. Iterate in
+  // reverse so that we don't have to update the argument indices as we go.
+  for (unsigned bbArgIndex : llvm::reverse(blockArgsToDrop)) {
+    Block &entryBlock = newOp.getRegion().front();
+    entryBlock.eraseArgument(bbArgIndex);
+  }
+
+  rewriter.replaceAllUsesWith(resultsToKeep, newOp.getResults());
+  rewriter.eraseOp(op);
+  return success();
+}
+
+struct DropUnusedGenericResult : public OpRewritePattern<IREE::PCF::GenericOp> {
+  using Base::Base;
+  LogicalResult matchAndRewrite(IREE::PCF::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    return dropUnusedResults(rewriter, genericOp);
+  }
+};
+
+struct DropUnusedLoopResult : public OpRewritePattern<IREE::PCF::LoopOp> {
+  using Base::Base;
+  LogicalResult matchAndRewrite(IREE::PCF::LoopOp loopOp,
+                                PatternRewriter &rewriter) const override {
+    return dropUnusedResults(rewriter, loopOp);
+  }
+};
+} // namespace
+
+void populatePCFDropUnusedResultPatterns(RewritePatternSet &patterns) {
+  patterns.add<DropUnusedGenericResult, DropUnusedLoopResult>(
```

Comment:
Why two patterns? They seem to be the same. Could just use a templated single pattern? Again, why cant we do this on the interface directly?